File: boj/solved/10815/Main.java
import java.io.OutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.StringTokenizer;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.InputStream;

public class Main {
    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        PrintWriter out = new PrintWriter(outputStream);
        Task solver = new Task();
        int testCount = 1;
        for (int i = 1; i <= testCount; i++) {
            solver.solve(i, in, out);
        }
        out.close();
    }
 
    static class Task {
        public void solve(int testNumber, InputReader in, PrintWriter out) {
            int n = in.nextInt();
            int[] a = new int[n];
            for (int i = 0; i < n; i++) {
                a[i] = in.nextInt();
            }
            Arrays.sort(a);
            int m = in.nextInt();
            for (int i = 0; i < m; i++) {
                int x = in.nextInt();
                int low = 0;
                int high = n - 1;
                while (low < high) {
                    int mid = (low + high) / 2;
                    if (a[mid] < x) {
                        low = mid + 1;
                    } else {
                        high = mid;
                    }
                }
                out.print((a[low] == x ? 1 : 0) + (i == m - 1 ? "\n" : " "));
            }
        }
    }
 
    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;
 
        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }
 
        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }
 
        public int nextInt() {
            return Integer.parseInt(next());
        }
    }
}

File: boj/solved/10815/Main.kt
import java.io.*
import java.util.*
 
object Main {
    @JvmStatic
    fun main(args: Array<String>) {
        val inputStream = System.`in`
        val outputStream: OutputStream = System.out
        val `in` = InputReader(inputStream)
        val out = PrintWriter(outputStream)
        val solver = Task()
        val testCount = 1
        for (i in 1..testCount) {
            solver.solve(i, `in`, out)
        }
        out.close()
    }
 
    internal class Task {
        fun solve(testNumber: Int, `in`: InputReader, out: PrintWriter) {
            var n = `in`.nextInt()
            var a = IntArray(n) { `in`.nextInt() }
            a.sort()
            var m = `in`.nextInt()
            var res = IntArray(m) { binarySearch(a, `in`.nextInt(), n) }
            out.println(res.joinToString(" "))
        }

        private fun binarySearch(a: IntArray, x: Int, n: Int): Int {
            var low = 0
            var high = n - 1
            while (low < high) {
                var mid = (low + high) / 2
                if (a[mid] < x) {
                    low = mid + 1
                } else {
                    high = mid
                }
            }
            return if (a[low] == x) 1 else 0
        }
    }
 
    internal class InputReader(stream: InputStream?) {
        var reader: BufferedReader
        var tokenizer: StringTokenizer?

        operator fun next(): String {
            while (tokenizer == null || !tokenizer!!.hasMoreTokens()) {
                tokenizer = try {
                    StringTokenizer(reader.readLine())
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }
            }
            return tokenizer!!.nextToken()
        }
 
        fun nextInt(): Int {
            return next().toInt()
        }
 
        init {
            reader = BufferedReader(InputStreamReader(stream), 32768)
            tokenizer = null
        }
    }
}